packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts

```diff
@@ -104,26 +104,55 @@ export const visualizeSharedCell: VisualizeSharedObject = async (
 };
 
 /**
- * Default {@link VisualizeSharedObject} for {@link DataObject}.
- * @remarks This takes in a `root` of type {@link ISharedDirectory} from {@link DataObject} and visualizes its children.
+ * Creates a visualizer function that captures the parent object's ID for use in generating the fluidObjectId.
+ * @param parentId - The ID of the parent DataObject to use in generating unique fluidObjectIds
+ * @returns A VisualizeSharedObject function that will use the parentId in its fluidObjectId generation
  */
-export const visualizeDataObject: VisualizeSharedObject = async (
-	dataObjectRoot: ISharedObject,
-	visualizeChildData: VisualizeChildData,
-): Promise<FluidObjectTreeNode> => {
-	const renderedChildData = (await visualizeSharedDirectory(
-		dataObjectRoot,
-		visualizeChildData,
-	)) as FluidObjectTreeNode; // TODO: Refactor the visualizer to accept generic type to avoid type casting.
+export function createDataObjectVisualizer(parentId: string): VisualizeSharedObject {
+	return async (
+		dataObjectRoot: ISharedObject,
+		visualizeChildData: VisualizeChildData,
+	): Promise<FluidObjectTreeNode> => {
+		const renderedChildData = (await visualizeSharedDirectory(
+			dataObjectRoot,
+			visualizeChildData,
+		)) as FluidObjectTreeNode;
 
-	return {
-		fluidObjectId: dataObjectRoot.id,
-		children: renderedChildData.children,
-		metadata: renderedChildData.metadata,
-		typeMetadata: "DataObject",
-		nodeKind: VisualNodeKind.FluidTreeNode,
+		return {
+			fluidObjectId: `${parentId}-${dataObjectRoot.id}`,
+			children: renderedChildData.children,
+			metadata: renderedChildData.metadata,
+			typeMetadata: "DataObject",
+			nodeKind: VisualNodeKind.FluidTreeNode,
+		};
 	};
-};
+}
+
+/**
+ * Creates a visualizer function that captures the parent object's ID for use in generating the fluidObjectId.
+ * @param parentId - The ID of the parent TreeDataObject to use in generating unique fluidObjectIds
+ * @returns A VisualizeSharedObject function that will use the parentId in its fluidObjectId generation
+ */
+export function createTreeDataObjectVisualizer(parentId: string): VisualizeSharedObject {
+	return async (
+		rootTree: ISharedObject,
+		visualizeChildData: VisualizeChildData,
+	): Promise<FluidObjectTreeNode> => {
+		const fluidObjectId = `${parentId}-${rootTree.id}`;
```

Comment:
Nit: in `createDataObjectVisualizer`, this is inlined into the return statement. Either way (inlined or variable) is fine, but I would prefer we be consistent :)